orttraining/orttraining/core/session/tensor_helper.cc

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/session/tensor_helper.h"
+#include "core/providers/cuda/cuda_common.h"
+
+namespace onnxruntime {
+namespace training {
+
+// Return the shape of a tensor slice.
+std::vector<int64_t> GetSliceShape(
+    const std::vector<int64_t>& shape,  // before-slicing tensor shape
+    const size_t& slice_axis,           // axis to slice along
+    const size_t& num_slices) {         // number of slices along the slicing axis
+  ORT_ENFORCE(shape.size() > 0);
+  ORT_ENFORCE(slice_axis < shape.size());
+  ORT_ENFORCE(num_slices > 0);
+
+  // Shape of slice along slice_axis.
+  std::vector<int64_t> slice_shape;
+
+  // Go through the original dimensions to get the dimensions after slicing.
+  for (size_t i_shape = 0; i_shape < shape.size(); ++i_shape) {
+    const auto d = shape[i_shape];
+
+    // Compute slice's shape.
+    if (i_shape == slice_axis) {
+      // Dimension shrinks due to slicing.
+      slice_shape.push_back(d / num_slices);
+    } else {
+      // Dimension not sliced, so we just copy its original value.
+      slice_shape.push_back(d);
+    }
+  }
+
+  return slice_shape;
+}
+
+// Given tensor's element type and shape, this function creates a tensor in the passed-in session.
+OrtValue CreateCpuTensorValue(
+    const MLDataType elem_type,
+    std::vector<int64_t> shape,
+    onnxruntime::InferenceSession& session_state) {
+  ORT_ENFORCE(elem_type->AsPrimitiveDataType(), "Tensor's element type must be a scalar type.");
+  ORT_ENFORCE(shape.size() > 0, "Shape vector must be non-empty.");
+
+  // Get CPU allocator from the session.
+  OrtMemoryInfo cpu_location(onnxruntime::CPU, OrtDeviceAllocator);
+  AllocatorPtr cpu_allocator = session_state.GetAllocator(cpu_location);
+
+  // Given a shape, allocate a tensor using CPU allocator.
+  auto cpu_tensor = onnxruntime::make_unique<Tensor>(elem_type, shape, cpu_allocator);
+
+  // Create type definition for the created tensor.
+  auto tensor_type = DataTypeImpl::GetType<Tensor>();
+
+  // Create OrtValue to wrap the allocated tensor.
+  OrtValue cpu_value{cpu_tensor.release(), tensor_type, tensor_type->GetDeleteFunc()};
+
+  return cpu_value;
+}
+
+// Copy a chunk of memory to CPU from GPU.
+void CopyGpuToCpu(
+    void* dst_ptr,
+    const void* src_ptr,
+    const size_t& size,
```

Comment:
>size_t& [](start = 10, length = 7)

nit: size_t type has the same size as a pointer. No need to pass it as a reference here.

Reply:
Thanks. I will change all function arguments of `const size_t&` to `const size.`

---
In reply to: [538734568](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538734568) [](ancestors = 538734568)